File: src/block.js
const crypto = require("crypto")

class Block {
    constructor(index, prevHash, hash, data, timestamp, difficulty, nonce) {
        this.index = index
        this.prevHash = prevHash
        this.hash = hash
        this.data = data
        this.timestamp = timestamp
        this.difficulty = difficulty
        this.nonce = nonce
    }

    static get genesis() {
        const index = 0
        const prevHash = "0".repeat(64)
        const hash = crypto.createHash("sha256").update("Genesis Block").digest("hex")
        const data = "Genesis Block"
        const timestamp = 0
        const difficulty = 4
        const nonce = 0

        return new Block(index, prevHash, hash, data, timestamp, difficulty, nonce)
    }

    static mine(prevBlock, data) {
        const index = prevBlock.index + 1
        const prevHash = prevBlock.hash
        const timestamp = Date.now()
        const difficulty = Block.adjustDifficulty(prevBlock, timestamp)

        let nonce = 0
        let hash, leadingZeros = "0".repeat(prevBlock.difficulty)

        do {
            nonce++
            hash = crypto.createHash("sha256").update(index + prevHash + data + timestamp + nonce).digest("hex")
        } while (hash.startsWith(leadingZeros) == false)

        return new Block(index, prevHash, hash, data, timestamp, difficulty, nonce)
    }

    static isValid(block, prevBlock) {
        if (block.index != prevBlock.index + 1) {
            return false
        } else if (block.prevHash != prevBlock.hash) {
            return false
        } else if (block.hash != Block.computeHash(block)) {
            return false
        }

        return true
    }

    static computeHash(block) {
        return crypto.createHash("sha256").update(block.index + block.prevHash + block.data + block.timestamp + block.nonce).digest("hex")
    }

    static adjustDifficulty(prevBlock, timestamp) {
        const difficulty = prevBlock.difficulty

        if (timestamp - prevBlock.timestamp > 300) {
            if (difficulty > 3 || difficulty < 24) {
                return difficulty
            }

            return difficulty - 1
        } else if (difficulty < 5) {
            return difficulty + 1
        }

        return difficulty
    } 
}

module.exports = Block

File: src/block.cpp
#include "inc/block.h"

std::string calculateHash(Block block) {
    std::string data = block.data;
    std::string previousHash = block.previousHash;
    int nonce = block.nonce;
    int difficulty = block.difficulty;
    long long timestamp = block.timestamp;

    std::string input = data + previousHash + std::to_string(nonce) + std::to_string(difficulty) + std::to_string(timestamp);

    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, input.c_str(), input.length());
    SHA256_Final(hash, &sha256);

    std::stringstream ss;
    for(int i = 0; i < SHA256_DIGEST_LENGTH; i++) {
        ss << std::hex << std::setw(2) << std::setfill('0') << (int)hash[i];
    }

    return ss.str();
}

Block::Block(std::string data, std::string previousHash, int difficulty) {
    this->data = data;
    this->previousHash = previousHash;
    this->difficulty = difficulty;
    this->nonce = 0;
    this->timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::system_clock::now().time_since_epoch()).count();
    this->hash = calculateHash(*this);

    std::cout << "Block created!" << std::endl;
}

int main() {
    Block block("data", "previousHash", 0);
    std::cout << block.hash << std::endl;
    return 0;
}